cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,365 @@
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	wait.Until(func() {
+		fc.actualRun(stopCh, fm, fwReconcileFrequency)
+	}, fwReconcileFrequency, stopCh)
+}
+
+func (fc *FirewallController) actualRun(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
+	defer cancel()
+
+	currentFirewall, err := fc.fwManager.Get(ctx)
+	if err != nil {
+		klog.Errorf("failed to get worker firewall: %s", err)
+	}
+
+	if currentFirewall != nil {
+		if fm.fwCache.isEqual(currentFirewall) {
+			return
+		}
+	}
+	fm.fwCache.updateCache(currentFirewall)
+	err = fc.ensureReconciledFirewall(ctx)
+	if err != nil {
+		klog.Errorf("failed to reconcile worker firewall: %s", err)
+	} else {
+		klog.Info("successfully reconciled firewall")
+	}
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	fw := fm.fwCache.getCachedFirewall()
+	if fw != nil {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fw.ID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("could not get firewall: %v", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %v", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+// Set will also reconcile away any changes to the outbound rules, name and tags.
+func (fm *firewallManagerOp) Set(ctx context.Context, fr *godo.FirewallRequest) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+
+	// A locally cached firewall exists, but the inbound rules don't match the expected
+	// service inbound rules. So we need to use the locally cached firewall ID to attempt
+	// to update the firewall APIs representation of the firewall with the new rules.
+	//
+	// Then we update the local cache with the firewall returned from the Update request.
```

Comment:
I'd either remove this comment, or move it right before we update the cache and explain why we only update our API operation succeeded (because otherwise, we wouldn't retry in the case of failures).